# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:
- **TASK 1:** Mark all launch sites on a map
- **TASK 2:** Mark the success/failed launches for each site on the map
- **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
import folium
import wget
import pandas as pd
from IPython.display import display

In [2]:
# Import folium MarkerCluster and MousePosition plugin
from folium.plugins import MarkerCluster, MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site. 


In [3]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df = pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [4]:
spacex_df.head()

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [6]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example, 


In [7]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle. 


Now, let's add a circle for each launch site in data frame `launch_sites`


_TODO:_  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [8]:
import folium
from folium.features import DivIcon

# Initialize the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Launch sites with coordinates
launch_sites = {
    'CCAFS_LC_40': [28.562302, -80.577356],
    'CCAFS_SLC_40': [28.563197, -80.576820],
    'KSC_LC_39A': [28.573255, -80.646895],
    'VAFB_SLC_4E': [34.632834, -120.610745],
}

# Add circles for each launch site
    

# Add markers for each launch site
for site, coordinates in launch_sites.items():
    circle = folium.Circle(coordinates, radius=1000, color='#d35400', fill=True).add_child(folium.Popup(site))
    site_map.add_child(circle)
    marker = folium.Marker(
        coordinates,
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html=f'<div style="font-size: 12; color:#d35400;"><b>{site}</b></div>'
        )
    )
    site_map.add_child(marker)

# Save the map as an HTML file
site_map.save("map_with_markers_and_circles.html")
display(site_map)


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:
- Are all launch sites in proximity to the Equator line?
- Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


- Based on these latitude values, all the launch sites are relatively close to the Equator line (closer to 0 degrees latitude). Generally, launch sites closer to the Equator provide certain advantages, such as increased rotational speed due to the Earth's rotation, which can be beneficial for rocket launches.

     CCAFS (Cape Canaveral Air Force Station) SLC-40 and CCAFS LC-40:
        These launch sites are associated with Cape Canaveral, which is located on the east coast of Florida. Therefore, it is likely that these launch sites are close to the coast.

    KSC LC-39A:
        This launch site is associated with Kennedy Space Center (KSC), which is also located on the east coast of Florida. Similar to CCAFS, it is likely to be close to the coast.

    VAFB SLC-4E:
        This launch site is associated with Vandenberg Air Force Base (VAFB), which is located on the west coast of California. Therefore, it is likely that this launch site is close to the coast.

Based on the general geographical locations of these launch sites, it is likely that they are in proximity to the coast. 

- Coastal launch sites provide safety, flexibility, and efficiency for space missions. Proximity to the ocean allows for versatile launch trajectories, enhances safety by directing rocket debris away from populated areas, and enables access to multiple orbital inclinations. Coastal locations also offer favorable weather conditions and logistical advantages.

# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [9]:
spacex_df.tail()

Launch Site        Lat      Long  class
51  CCAFS SLC-40  28.563197 -80.57682      0
52  CCAFS SLC-40  28.563197 -80.57682      0
53  CCAFS SLC-40  28.563197 -80.57682      0
54  CCAFS SLC-40  28.563197 -80.57682      1
55  CCAFS SLC-40  28.563197 -80.57682      0

Next, let's create markers for all launch records. 
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


_TODO:_ Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [10]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [11]:
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)

In [12]:
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

_TODO:_ For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [13]:
# Create a MarkerCluster layer
marker_cluster = MarkerCluster().add_to(site_map)

# Iterate through DataFrame and add markers and circles
for index, record in spacex_df.iterrows():
    latitude = record['Lat']
    longitude = record['Long']

    # Set custom icon color based on 'marker_color' column
    icon_color = record['marker_color']
    
    # Create a Circle for each location
    circle = folium.Circle(location=[latitude, longitude], color='#d35400', fill=True).add_child(folium.Popup(record['Launch Site']))

    # Create a Marker for each location with a custom icon color
    marker = folium.Marker(location=[latitude, longitude],
                           popup=record['Launch Site'],
                           icon=folium.Icon(color='white', icon_color=icon_color,), 
                           tooltip=f'<div style="font-size: 12; color:#d35400;"><b>{record["Launch Site"]}</b></div>')
                           
                            
    # Add the Circle to the map
    circle.add_to(site_map)
    # Add the Marker to the MarkerCluster
    marker.add_to(marker_cluster)

# Save the map to an HTML file or display it
display(site_map)
site_map.save("spacex_map_with_clusters.html")


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [14]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [15]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

_TODO:_ Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [16]:
CCAFS_SLC_40_lat = 28.56319
CCAFS_SLC_40_long = -80.57682
coastline_lat = 28.56393 
coastline_long = -80.56783

distance_coastline = calculate_distance(CCAFS_SLC_40_lat, CCAFS_SLC_40_long, coastline_lat, coastline_long)
distance_coastline

0.8820974219701957

In [17]:
# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

_TODO:_ After obtained its coordinate, create a `folium.Marker` to show the distance


In [18]:
CCAFS_SLC_40_coordinate = [CCAFS_SLC_40_lat, CCAFS_SLC_40_long] 
coastline_coordinate = [coastline_lat, coastline_long] 
distance_marker = folium.Marker(
    coastline_coordinate,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline),
        )
    )
site_map.add_child(distance_marker)
site_map

In [19]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

_TODO:_ Draw a `PolyLine` between a launch site to the selected coastline point


In [20]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate



lines=folium.PolyLine(locations=[CCAFS_SLC_40_coordinate, coastline_coordinate], weight=1)
site_map.add_child(lines)

Your updated map with distance line should look like the following screenshot:


_TODO:_ Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [21]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
CCAFS_SLC_40_lat = 28.56341
CCAFS_SLC_40_long = -80.57679

highway_lat = 28.56409
highway_long = -80.57092
highway_coordinate = [highway_lat, highway_long]

railway_lat = 28.56447
railway_long = -80.58676
railway_coordinate = [railway_lat, railway_long]

city_lat = 28.10309
city_long = -80.6351
city_coordinate = [city_lat, city_long]

distance_highway = calculate_distance(CCAFS_SLC_40_lat,CCAFS_SLC_40_long, highway_lat, highway_long)
distance_railway = calculate_distance(CCAFS_SLC_40_lat, CCAFS_SLC_40_long, railway_lat, railway_long)
distance_city = calculate_distance(CCAFS_SLC_40_lat, CCAFS_SLC_40_long, city_lat, city_long)


distance_marker_railway = folium.Marker(
    railway_coordinate,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_railway),
        )
    )

distance_marker_city = folium.Marker(
    city_coordinate,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_city),
        )
    )

distance_marker_highway = folium.Marker(
    highway_coordinate,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_highway),
        )
    )

lines_city=folium.PolyLine(locations=[CCAFS_SLC_40_coordinate, city_coordinate], weight=1)

lines_highway=folium.PolyLine(locations=[CCAFS_SLC_40_coordinate, highway_coordinate], weight=1)

lines_railway=folium.PolyLine(locations=[CCAFS_SLC_40_coordinate, railway_coordinate], weight=1)


site_map.add_child(lines_railway)
site_map.add_child(lines_highway)
site_map.add_child(lines_city)


site_map.add_child(distance_marker_railway)
site_map.add_child(distance_marker_city)
site_map.add_child(distance_marker_highway)



site_map.save("spacex_map_with_clusters.html")
site_map

After you plot distance lines to the proximities, you can answer the following questions easily:
- Are launch sites in close proximity to railways?
- Are launch sites in close proximity to highways?
- Are launch sites in close proximity to coastline?
- Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


- launch sites in close proximity to railways which is 0.98 km, highways which is 0.58, coastline which is 0.88 km and far from cities which is approx 51.52 km.

So Selecting optimal launch sites involves criteria like close proximity to railways and highways for logistics, proximity to coastlines for safe launches over water, and being far from cities to ensure public safety. Balancing these factors optimizes transportation and minimizes risks during space launches.
